File: sdk/go/canonical_test.go
package hardenhmac

import (
	"testing"
)

func TestBuildCanonicalString_BasicGet(t *testing.T) {
	result, err := BuildCanonicalString("GET", "/api/users", "", 1700000000, nil, nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	expected := "GET\n/api/users\n\n\n1700000000"
	if result != expected {
		t.Errorf("basic GET:\ngot:  %q\nwant: %q", result, expected)
	}
}

func TestBuildCanonicalString_PostWithBody(t *testing.T) {
	result, err := BuildCanonicalString("POST", "/api/users", `{"name":"Alice","email":"alice@example.com"}`, 1700000001, nil, nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	expected := "POST\n/api/users\n\n{\"name\":\"Alice\",\"email\":\"alice@example.com\"}\n1700000001"
	if result != expected {
		t.Errorf("POST with body:\ngot:  %q\nwant: %q", result, expected)
	}
}

func TestBuildCanonicalString_MethodUppercased(t *testing.T) {
	result, err := BuildCanonicalString("gEt", "/api/mixed-case", "", 1700000020, nil, nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	expected := "GET\n/api/mixed-case\n\n\n1700000020"
	if result != expected {
		t.Errorf("mixed case method:\ngot:  %q\nwant: %q", result, expected)
	}
}

func TestBuildCanonicalString_SignedHeaders_AuthAndX(t *testing.T) {
	config := &SignedHeadersConfig{
		IncludeAuthorization: true,
		IncludeXHeaders:      true,
	}
	headers := map[string]string{
		"Authorization":      "Bearer eyJhbGciOiJIUzI1NiJ9.test",
		"X-Request-Id":       "req-abc-123",
		"Content-Type":       "application/json",
		"X-Harden-Timestamp": "1700000004",
	}
	result, err := BuildCanonicalString("POST", "/api/orders", `{"item":"widget","qty":5}`, 1700000004, config, headers)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	expected := "POST\n/api/orders\nauthorization:Bearer eyJhbGciOiJIUzI1NiJ9.test\nx-request-id:req-abc-123\n{\"item\":\"widget\",\"qty\":5}\n1700000004"
	if result != expected {
		t.Errorf("signed headers auth+x:\ngot:  %q\nwant: %q", result, expected)
	}
}

func TestBuildCanonicalString_MultipleSortedHeaders(t *testing.T) {
	config := &SignedHeadersConfig{
		IncludeAuthorization: true,
		IncludeXHeaders:      true,
		AdditionalHeaders:    []string{"Content-Type"},
	}
	headers := map[string]string{
		"Authorization":       "Bearer token123",
		"Content-Type":        "application/json",
		"X-Custom-A":          "value-a",
		"X-Custom-B":          "value-b",
		"Accept":              "application/json",
		"X-Harden-Signature":  "will-be-excluded",
	}
	result, err := BuildCanonicalString("POST", "/api/data", `{"key":"value"}`, 1700000007, config, headers)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	expected := "POST\n/api/data\nauthorization:Bearer token123\ncontent-type:application/json\nx-custom-a:value-a\nx-custom-b:value-b\n{\"key\":\"value\"}\n1700000007"
	if result != expected {
		t.Errorf("multiple sorted headers:\ngot:  %q\nwant: %q", result, expected)
	}
}

func TestBuildCanonicalString_ExcludeHeaders(t *testing.T) {
	config := &SignedHeadersConfig{
		IncludeAuthorization: true,
		IncludeXHeaders:      true,
		ExcludeHeaders:       []string{"X-Custom-B"},
	}
	headers := map[string]string{
		"Authorization": "Bearer excluded-test",
		"X-Custom-A":    "keep-this",
		"X-Custom-B":    "exclude-this",
	}
	result, err := BuildCanonicalString("GET", "/api/filtered", "", 1700000009, config, headers)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	expected := "GET\n/api/filtered\nauthorization:Bearer excluded-test\nx-custom-a:keep-this\n\n1700000009"
	if result != expected {
		t.Errorf("exclude headers:\ngot:  %q\nwant: %q", result, expected)
	}
}

func TestBuildCanonicalString_HeaderValueTrimming(t *testing.T) {
	config := &SignedHeadersConfig{
		IncludeAuthorization: true,
		IncludeXHeaders:      true,
	}
	headers := map[string]string{
		"Authorization": "  Bearer spaced-token  ",
		"X-Padded":      "  trimmed-value  ",
	}
	result, err := BuildCanonicalString("GET", "/api/trim-test", "", 1700000010, config, headers)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	expected := "GET\n/api/trim-test\nauthorization:Bearer spaced-token\nx-padded:trimmed-value\n\n1700000010"
	if result != expected {
		t.Errorf("header trimming:\ngot:  %q\nwant: %q", result, expected)
	}
}

func TestBuildCanonicalString_EmptyHeaderValue(t *testing.T) {
	config := &SignedHeadersConfig{
		IncludeXHeaders: true,
	}
	headers := map[string]string{
		"X-Empty": "",
	}
	result, err := BuildCanonicalString("GET", "/api/empty-header", "", 1700000021, config, headers)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	expected := "GET\n/api/empty-header\nx-empty:\n\n1700000021"
	if result != expected {
		t.Errorf("empty header value:\ngot:  %q\nwant: %q", result, expected)
	}
}

func TestBuildCanonicalString_AuthExcludedWhenDisabled(t *testing.T) {
	config := &SignedHeadersConfig{
		IncludeAuthorization: false,
		IncludeXHeaders:      true,
	}
	headers := map[string]string{
		"Authorization": "Bearer should-not-be-signed",
		"X-Request-Id":  "req-999",
	}
	result, err := BuildCanonicalString("GET", "/api/auth-excluded", "", 1700000022, config, headers)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	expected := "GET\n/api/auth-excluded\nx-request-id:req-999\n\n1700000022"
	if result != expected {
		t.Errorf("auth excluded:\ngot:  %q\nwant: %q", result, expected)
	}
}

func TestBuildCanonicalString_RootPath(t *testing.T) {
	result, err := BuildCanonicalString("GET", "/", "", 1700000011, nil, nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	expected := "GET\n/\n\n\n1700000011"
	if result != expected {
		t.Errorf("root path:\ngot:  %q\nwant: %q", result, expected)
	}
}

func TestBuildCanonicalString_QueryParams(t *testing.T) {
	result, err := BuildCanonicalString("PUT", "/api/users/123?active=true&role=admin", `{"name":"Bob"}`, 1700000002, nil, nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	expected := "PUT\n/api/users/123?active=true&role=admin\n\n{\"name\":\"Bob\"}\n1700000002"
	if result != expected {
		t.Errorf("query params:\ngot:  %q\nwant: %q", result, expected)
	}
}

func TestBuildSignedHeaders_ReturnsNames(t *testing.T) {
	config := SignedHeadersConfig{
		IncludeAuthorization: true,
		IncludeXHeaders:      true,
	}
	headers := map[string]string{
		"Authorization": "Bearer token",
		"X-Request-Id":  "abc",
	}
	str, names := BuildSignedHeaders(config, headers)
	if str != "authorization:Bearer token\nx-request-id:abc" {
		t.Errorf("header string: got %q", str)
	}
	if len(names) != 2 || names[0] != "authorization" || names[1] != "x-request-id" {
		t.Errorf("header names: got %v", names)
	}
}

func TestBuildCanonicalString_NilConfig(t *testing.T) {
	// nil config should produce empty signed headers
	result, err := BuildCanonicalString("GET", "/test", "", 1000, nil, map[string]string{
		"Authorization": "Bearer tok",
		"X-Foo":         "bar",
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	expected := "GET\n/test\n\n\n1000"
	if result != expected {
		t.Errorf("nil config:\ngot:  %q\nwant: %q", result, expected)
	}
}

// ATK-1: Newline injection validation
func TestBuildCanonicalString_NewlineInMethod_ReturnsError(t *testing.T) {
	_, err := BuildCanonicalString("GET\nX-Injected:evil", "/api/test", "", 1700000000, nil, nil)
	if err == nil {
		t.Fatal("expected error for newline in method, got nil")
	}
	if err.Error() != "method must not contain newline characters" {
		t.Errorf("unexpected error message: %q", err.Error())
	}
}

func TestBuildCanonicalString_NewlineInPath_ReturnsError(t *testing.T) {
	_, err := BuildCanonicalString("GET", "/api/test\nX-Injected:evil", "", 1700000000, nil, nil)
	if err == nil {
		t.Fatal("expected error for newline in path, got nil")
	}
	if err.Error() != "path must not contain newline characters" {
		t.Errorf("unexpected error message: %q", err.Error())
	}
}

// ATK-3: X-Harden-Client-Id always signed
func TestBuildCanonicalString_ClientIdSignedWithoutXHeaders(t *testing.T) {
	config := &SignedHeadersConfig{
		IncludeAuthorization: false,
		IncludeXHeaders:      false,
	}
	headers := map[string]string{
		"X-Harden-Client-Id": "my-client",
		"X-Custom-Header":    "should-not-be-signed",
	}
	result, err := BuildCanonicalString("GET", "/api/test", "", 1700000000, config, headers)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	expected := "GET\n/api/test\nx-harden-client-id:my-client\n\n1700000000"
	if result != expected {
		t.Errorf("client-id without x-headers:\ngot:  %q\nwant: %q", result, expected)
	}
}

File: sdk/typescript/packages/hmac/tests/canonical.test.ts
import { describe, it, expect } from "vitest";
import { buildCanonicalString, buildSignedHeaders } from "../src/canonical.js";
import type { SignedHeadersConfig } from "../src/config.js";

const noneConfig: SignedHeadersConfig = {
  includeAuthorization: false,
  includeXHeaders: false,
  additionalHeaders: [],
  excludeHeaders: [],
};

describe("buildCanonicalString", () => {
  it("builds basic GET canonical string", () => {
    const result = buildCanonicalString({
      method: "GET",
      path: "/api/users",
      body: "",
      timestamp: 1700000000,
      signedHeadersConfig: noneConfig,
    });
    expect(result).toBe("GET\n/api/users\n\n\n1700000000");
  });

  it("includes body in POST", () => {
    const body = '{"name":"Alice"}';
    const result = buildCanonicalString({
      method: "POST",
      path: "/api/users",
      body,
      timestamp: 1700000001,
      signedHeadersConfig: noneConfig,
    });
    expect(result).toBe(`POST\n/api/users\n\n${body}\n1700000001`);
  });

  it("uppercases method", () => {
    const result = buildCanonicalString({
      method: "get",
      path: "/api/users",
      body: "",
      timestamp: 1700000000,
      signedHeadersConfig: noneConfig,
    });
    expect(result).toMatch(/^GET\n/);
  });

  it("includes query string in path", () => {
    const result = buildCanonicalString({
      method: "GET",
      path: "/api/users?active=true",
      body: "",
      timestamp: 1700000000,
      signedHeadersConfig: noneConfig,
    });
    expect(result).toContain("/api/users?active=true");
  });

  it("handles root path", () => {
    const result = buildCanonicalString({
      method: "GET",
      path: "/",
      body: "",
      timestamp: 1700000000,
      signedHeadersConfig: noneConfig,
    });
    expect(result).toBe("GET\n/\n\n\n1700000000");
  });

  it("sorts signed headers alphabetically", () => {
    const config: SignedHeadersConfig = {
      includeAuthorization: true,
      includeXHeaders: true,
      additionalHeaders: [],
      excludeHeaders: [],
    };
    const result = buildCanonicalString({
      method: "GET",
      path: "/",
      body: "",
      timestamp: 1700000000,
      signedHeadersConfig: config,
      requestHeaders: {
        "X-Request-Id": "abc",
        Authorization: "Bearer tok",
      },
    });
    expect(result).toContain(
      "authorization:Bearer tok\nx-request-id:abc"
    );
  });

  it("excludes X-Harden-* headers", () => {
    const config: SignedHeadersConfig = {
      includeAuthorization: false,
      includeXHeaders: true,
      additionalHeaders: [],
      excludeHeaders: [],
    };
    const result = buildCanonicalString({
      method: "GET",
      path: "/",
      body: "",
      timestamp: 1700000000,
      signedHeadersConfig: config,
      requestHeaders: {
        "X-Request-Id": "abc",
        "X-Harden-Signature": "should-be-excluded",
      },
    });
    expect(result).toContain("x-request-id:abc");
    expect(result).not.toContain("x-harden-");
  });

  it("applies exclude override", () => {
    const config: SignedHeadersConfig = {
      includeAuthorization: false,
      includeXHeaders: true,
      additionalHeaders: [],
      excludeHeaders: ["X-Custom-B"],
    };
    const result = buildCanonicalString({
      method: "GET",
      path: "/",
      body: "",
      timestamp: 1700000000,
      signedHeadersConfig: config,
      requestHeaders: {
        "X-Custom-A": "keep",
        "X-Custom-B": "exclude",
      },
    });
    expect(result).toContain("x-custom-a:keep");
    expect(result).not.toContain("x-custom-b");
  });

  it("trims header values", () => {
    const config: SignedHeadersConfig = {
      includeAuthorization: true,
      includeXHeaders: false,
      additionalHeaders: [],
      excludeHeaders: [],
    };
    const result = buildCanonicalString({
      method: "GET",
      path: "/",
      body: "",
      timestamp: 1700000000,
      signedHeadersConfig: config,
      requestHeaders: {
        Authorization: "  Bearer tok  ",
      },
    });
    expect(result).toContain("authorization:Bearer tok");
  });

  // ATK-1: Newline injection validation
  it("throws on newline in method", () => {
    expect(() =>
      buildCanonicalString({
        method: "GET\nX-Injected:evil",
        path: "/api/test",
        body: "",
        timestamp: 1700000000,
        signedHeadersConfig: noneConfig,
      })
    ).toThrow("method must not contain newline characters");
  });

  it("throws on newline in path", () => {
    expect(() =>
      buildCanonicalString({
        method: "GET",
        path: "/api/test\nX-Injected:evil",
        body: "",
        timestamp: 1700000000,
        signedHeadersConfig: noneConfig,
      })
    ).toThrow("path must not contain newline characters");
  });

  // ATK-3: X-Harden-Client-Id always signed
  it("signs X-Harden-Client-Id even when includeXHeaders is false", () => {
    const config: SignedHeadersConfig = {
      includeAuthorization: false,
      includeXHeaders: false,
      additionalHeaders: [],
      excludeHeaders: [],
    };
    const result = buildCanonicalString({
      method: "GET",
      path: "/api/secure",
      body: "",
      timestamp: 1700000050,
      signedHeadersConfig: config,
      requestHeaders: {
        "X-Harden-Client-Id": "my-client",
        "X-Custom-Header": "should-not-be-signed",
      },
    });
    expect(result).toContain("x-harden-client-id:my-client");
    expect(result).not.toContain("x-custom-header");
  });

  it("includes additional headers", () => {
    const config: SignedHeadersConfig = {
      includeAuthorization: false,
      includeXHeaders: false,
      additionalHeaders: ["Content-Type"],
      excludeHeaders: [],
    };
    const result = buildCanonicalString({
      method: "GET",
      path: "/",
      body: "",
      timestamp: 1700000000,
      signedHeadersConfig: config,
      requestHeaders: {
        "Content-Type": "application/json",
        Accept: "text/html",
      },
    });
    expect(result).toContain("content-type:application/json");
    expect(result).not.toContain("accept");
  });
});

describe("buildSignedHeaders", () => {
  it("returns header names sorted", () => {
    const config: SignedHeadersConfig = {
      includeAuthorization: true,
      includeXHeaders: true,
      additionalHeaders: [],
      excludeHeaders: [],
    };
    const { headerString, headerNames } = buildSignedHeaders(config, {
      "X-Request-Id": "abc",
      Authorization: "Bearer tok",
    });
    expect(headerNames).toEqual(["authorization", "x-request-id"]);
    expect(headerString).toBe(
      "authorization:Bearer tok\nx-request-id:abc"
    );
  });
});
